docs/TIPS.md

```diff
@@ -0,0 +1,251 @@
+
+Unfortunately, running a subprocess inside GDB can magically cause the crash to disappear and appear in another subprocess. Trying to chase the crash can be quite time-consuming and it is often more effective to attach GDB to every subprocess. To avoid running `run` manually in GDB on every subprocess, the following options are useful: `--eval-command=run --eval-command="set disable-randomization on" --eval-command=quit`. This will automatically start the program inside GDB and quit unless a crash causes GDB to pause.
+
+Sometimes, the offending subprocess is created using the Python subprocess libraries rather than from a string run as a shell script. For example, in `test/integration/virtualenv/virtualenv.py` in the function `install_wheel`:
+
+```
+    cmd = [
+        py_executable, '-c',
+        'import sys, pip; sys.exit(pip.main(["install", "--ignore-installed"] + sys.argv[1:]))',
+    ] + project_names
+    ...
+    try:
+        call_subprocess(cmd, show_stdout=False,
+            extra_env = {
+                'PYTHONPATH': pythonpath,
+                'PIP_FIND_LINKS': findlinks,
+                'PIP_USE_WHEEL': '1',
+                'PIP_PRE': '1',
+                'PIP_NO_INDEX': '1'
+            }
+        )
+    finally:
+        ...
+```
+
+You will want to wrap the command inside GDB
+
+```
+    cmd = [
+        "gdb", "--args",
+        py_executable, '-c',
+        'import sys, pip; sys.exit(pip.main(["install", "--ignore-installed"] + sys.argv[1:]))',
+    ] + project_names
+```
+
+and make sure that you can see the GDB prompt by changing `show_stdout=False` to `show_stdout=True`.
+
+##### Problem: I still can't isolate the crashing subprocess
+
+Wrapping a subprocess call in GDB can be quick and effective, but sometimes you run into a [heisenbug](https://en.wikipedia.org/wiki/Heisenbug) and GDB hides the problem no matter what you do.
+
+If you can isolate the nature of the crash (e.g. segmentation fault), you may be able to add a signal handler with an infinite loop and attach GDB to the process after-the-fact by starting GDB and using `attach PID` where `PID` is the hanging `pyston_dbg` process id. You may want to look inside `src/codegen/entry.cpp` for signal handlers. If you want to pause on segfault, you may want to add a snippet of code like:
+
+```
+static void handle_sigsegv(int signum) {
+
+    assert(signum == SIGSEGV);
+
+    fprintf(stderr, "SIGSEGV, printing stack trace\n");
+
+    _printStacktrace();
+
+    printf("HANDLE SIGSEGV %d\n", getpid());
+
+    fprintf(stderr, "HANDLE SIGSEGV %d\n", getpid());
+
+    while (true) {
+    }
+
+}
+
+void initCodegen() {
+    ...
+    signal(SIGSEGV, &handle_sigsegv);
+    ...
+}
+```
```

Comment:
Could be worth mentioning `PYSTON_RUN_ARGS=G` and `=P`.  `-P` should probably do the segfault thing you mention here :)